src/cycles/incremental.jl

```diff
@@ -0,0 +1,266 @@
+the ict intact. If not, update the underlying graph and return true.
+
+# Optional `f!` Argument
+
+By default the `add_edge!` function is used to update the underlying graph.
+However, for more complicated graphs, users may wish to manually specify the
+graph update operation. This may be accomplished by passing the optional `f!`
+callback arhgument. This callback is called on the underlying graph when no
+cycle is detected and is required to modify the underlying graph in order to
+effectuate the proposed edge addition.
+
+# Batched edge additions
+
+Optionally, either `v` or `w` (depending on the `in_out_reverse` flag) may be a
+collection of vertices representing a batched addition of vertices sharing a
+common source or target more efficiently than individual updates.
+
+## Example
+
+```jldoctest
+julia> G = SimpleDiGraph(3)
+
+julia> ict = IncrementalCycleTracker(G)
+BFGT_N cycle tracker on {3, 0} directed simple Int64 graph
+
+julia> add_edge_checked!(ict, 1, 2)
+true
+
+julia> collect(edges(G))
+1-element Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}:
+ Edge 1 => 2
+
+julia> add_edge_checked!(ict, 2, 3)
+true
+
+julia> collect(edges(G))
+2-element Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}:
+ Edge 1 => 2
+ Edge 2 => 3
+
+julia> add_edge_checked!(ict, 3, 1) # Would add a cycle
+false
+
+julia> collect(edges(G))
+2-element Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}:
+Edge 1 => 2
+Edge 2 => 3
+```
+"""
+function add_edge_checked! end
+
+to_edges(v::Integer, w::Integer) = (v=>w,)
+to_edges(v::Integer, ws) = zip(repeated(v), ws)
+to_edges(vs, w::Integer) = zip(vs, repeated(w))
+
+add_edge_checked!(ict::IncrementalCycleTracker, vs, ws) = add_edge_checked!(ict, vs, ws) do g
+    foreach(((v, w),)->add_edge!(g, v, w), to_edges(vs, ws))
+end
+
+# Utilities
+"""
+    struct TransactionalVector
+
+A vector with one checkpoint that may be reverted to by calling `revert!`. The setpoint itself
+is set by calling `commit!`.
+"""
+struct TransactionalVector{T} <: AbstractVector{T}
+    v::Vector{T}
+    log::Vector{Pair{Int, T}}
+    TransactionalVector(v::Vector{T}) where {T} =
+        new{T}(v, Vector{Pair{Int, T}}())
+end
+
+function commit!(v::TransactionalVector)
+    empty!(v.log)
+    return nothing
+end
+
+function revert!(vec::TransactionalVector)
+    for (idx, val) in reverse(vec.log)
+        vec.v[idx] = val
+    end
+    return nothing
+end
+
+function Base.setindex!(vec::TransactionalVector, val, idx)
+    oldval = vec.v[idx]
+    vec.v[idx] = val
+    push!(vec.log, idx=>oldval)
+    return nothing
+end
+Base.getindex(vec::TransactionalVector, idx) = vec.v[idx]
+Base.size(vec::TransactionalVector) = size(vec.v)
+
+function weak_topological_levels(g::AbstractGraph{I}) where {I}
+    levels = fill(-1, nv(g))
+    # Fill topological levels via dynamic programming dfs
+    worklist = Vector{Int}()
+    all_verts_max_level = 0
+    for v in vertices(g)
+        push!(worklist, v)
```

Comment:
```suggestion
        levels[v] != -1 && continue
        push!(worklist, v)
```
If I'm not mistaking, this can save a bunch of `worklist` allocations and iterations to `outneighbors`.